File: 2023/day_4/second_star/haskell/app/Main.hs
import Data.List.Split (splitOn)
import Data.List (intersect)

main :: IO ()
main = do
    content <- readFile "../../input.txt"
    let cards = map parseCard $ lines content
    let copies = replicate (length cards) 1
    let finalCopies = foldl processCard copies cards
    print $ sum finalCopies

parseCard :: String -> (Int, Int)
parseCard card = 
    let [cardNumber, numbers] = splitOn ":" card
        [winning, mine] = splitOn "|" numbers
        winningNumbers = map read $ words winning :: [Int]
        myNumbers = map read $ words mine :: [Int]
        matches = length $ intersect myNumbers winningNumbers
        cardNumberInt = read (last (words cardNumber)) :: Int
    in (cardNumberInt, matches)

processCard :: [Int] -> (Int, Int) -> [Int]
processCard copies (cardNumber, matches) = 
    let copiesOfCard = copies !! (cardNumber - 1)
    in [if i >= cardNumber && i < cardNumber + matches then c + copiesOfCard else c | (c, i) <- zip copies [0..]]

File: 2023/day_4/second_star/rust/src/main.rs
fn main() {
    let input = "../../input.txt";
    let input_file = std::fs::read_to_string(input).expect("Error reading input file");
    let cards = input_file
        .lines()
        .map(|line| {
            let (card, nums) = line.split_once(':').unwrap();
            let card: i32 = card.split(' ').last().unwrap().parse().unwrap();
            let (win, my) = nums.split_once('|').unwrap();
            let win = win
                .trim()
                .split(' ')
                .filter(|n| n.trim() != "")
                .map(|n| n.parse::<i32>().unwrap())
                .collect::<Vec<_>>();
            let my = my
                .trim()
                .split(' ')
                .filter(|n| n.trim() != "")
                .map(|n| n.parse::<i32>().expect(&format!("NaN {n} Nan")))
                .collect::<Vec<_>>();
            let winning = win.iter().filter(|n| my.contains(n)).count();
            (card, winning)
        })
        .collect::<Vec<_>>();
    let mut copies = vec![1; cards.len()];
    for i in 0..cards.len() {
        let (card, win) = cards[i];
        for j in 0..win {
            copies[j+card as usize] += copies[card as usize-1];
        }
    }
    println!("{}", copies.iter().sum::<i32>());
}